easybuild/easyblocks/o/openssl_wrapper.py

```diff
@@ -0,0 +1,269 @@
+            self.ssl_sysbin = which('openssl11')
+
+        if not self.ssl_sysbin:
+            self.ssl_sysbin = which(self.name.lower())
+
+    def fetch_step(self, *args, **kwargs):
+        """Fetch sources if OpenSSL component is needed"""
+        if not all([self.ssl_syslib, self.ssl_sysheader]):
+            super(EB_OpenSSL_wrapper, self).fetch_step(*args, **kwargs)
+
+    def extract_step(self):
+        """Extract sources if OpenSSL component is needed"""
+        if not all([self.ssl_syslib, self.ssl_sysheader]):
+            super(EB_OpenSSL_wrapper, self).extract_step()
+
+    def install_step(self):
+        """Symlink target OpenSSL installation"""
+
+        if self.ssl_syslib and self.ssl_sysheader:
+            # note: symlink to individual files, not directories,
+            # since directory symlinks get resolved easily...
+
+            # link OpenSSL libraries in system
+            lib64_dir = os.path.join(self.installdir, 'lib64')
+            lib64_engines_dir = os.path.join(lib64_dir, os.path.basename(self.ssl_sysengines))
+            mkdir(lib64_engines_dir, parents=True)
+
+            # link existing known libraries
+            ssl_syslibdir = os.path.dirname(self.ssl_syslib)
+            lib_files = [os.path.join(ssl_syslibdir, x) for x in self.openssl_libs]
+            for libso in lib_files:
+                symlink(libso, os.path.join(lib64_dir, os.path.basename(libso)))
+
+            # link engines library files
+            lib64_engines_dir = os.path.join(lib64_dir, os.path.basename(self.ssl_sysengines))
+            engine_lib_pattern = [os.path.join(self.ssl_sysengines, '*')]
+            for engine_lib in expand_glob_paths(engine_lib_pattern):
+                symlink(engine_lib, os.path.join(lib64_engines_dir, os.path.basename(engine_lib)))
+
+            # relative symlink for unversioned libraries
+            cwd = change_dir(lib64_dir)
+            for libso in self.openssl_libs:
+                unversioned_lib = '%s.%s' % (libso.split('.')[0], get_shared_lib_ext())
+                symlink(libso, unversioned_lib, use_abspath_source=False)
+            change_dir(cwd)
+
+            # link OpenSSL headers in system
+            include_dir = os.path.join(self.installdir, 'include', self.name.lower())
+            mkdir(include_dir, parents=True)
+            include_pattern = [os.path.join(self.ssl_sysheader, '*')]
+            for header_file in expand_glob_paths(include_pattern):
+                symlink(header_file, os.path.join(include_dir, os.path.basename(header_file)))
+
+            # link OpenSSL binary in system
+            bin_dir = os.path.join(self.installdir, 'bin')
+            mkdir(bin_dir)
+            symlink(self.ssl_sysbin, os.path.join(bin_dir, self.name.lower()))
+
+        else:
+            # without wrapping, reduce matrix of library names to the first option in each family
+            self.openssl_libs = [lib_name[0] for lib_name in self.openssl_libs]
```

Comment:
There's a problem here, when OpenSSL 1.0 gets installed from source on Ubuntu 18.04, the sanity check fails:

```
== sanity checking...
  >> file 'bin/openssl' found: OK
  >> file 'lib/l' found: FAILED
  >> file 'lib/l' found: FAILED
  >> file 'lib/l.so' found: FAILED
  >> file 'lib/l.so' found: FAILED
  >> (non-empty) directory 'include/openssl' found: OK
  >> (non-empty) directory 'lib/engines' found: OK
  >> running command 'ssl_ver=$(openssl version); [ ${ssl_ver:8:3} == '1.0' ]' ...
  >> result for command 'ssl_ver=$(openssl version); [ ${ssl_ver:8:3} == '1.0' ]': OK
```

The (really) strange thing is that I'm not seeing this on other systems where `OpenSSL/1.0` is being built from source (like on RHEL8 and Ubuntu 20.04).

Could be related to using Python 2 vs Python 3 somehow?!

When I log `self.openssl_libs` right after this statement, I get:
```
== 2021-05-19 07:45:33,411 openssl_wrapper.py:244 INFO self.openssl_libs: ['l', 'l']
```

Reply:
I see what happens, my fault...

Reply:
Before this statement, we have:
```
== 2021-05-19 07:50:49,663 openssl_wrapper.py:245 INFO self.openssl_libs: ['libssl.so.1.0.0', 'libcrypto.so.1.0.0']
```

Reply:
The problem was that the dimension of `self.openssl_libs` was being reduced twice in the case where the host system had the libraries of OpenSSL but **not** the headers. I guess that your Ubuntu 18.04 lacks `openssl-dev`.